### ISSUE TITLE
bounds checking and error handling

### ISSUE THREAD

**Issue by psahabu** [chapel-lang/chapel#6629]
## motivation
It may be necessary for array accesses to throw errors when indexing out of bounds. However, doing so with the current implementation adds overhead for both error and non-error cases due to conditionals inserted by the lowering of error handling:

``` chapel
// required to handle errors
try {
  maybeOutOfBounds();
} catch {
  handleOOB();
}

// lowered
var e: Error;
maybeOutOfBounds(e);
if e then
  handleOOB();
```

Doing these checks on every access would cause an unacceptable loss of performance, so it is necessary for us to have some way to remove those conditional checks that is consistent with the current error handling system.

### status quo
proposed by @mppf 

Currently `--fast` produces undefined behavior while `--checks` produces a `halt()`. The idea that undefined behavior should be paired with a `throw` is strange on face because the program crashes in one instance while recovering in the other. Hence, the current situation should be maintained.

### unchecked throws
proposed by @mppf 

Mark a few baseline functions as _unchecked throws_, meaning that they are technically capable of throwing but that it's unreasonable for Chapel programmers to always write code that can handle their errors (out of bounds, out of memory, etc.). If a function does not throw but calls an _unchecked throws_ function without fully handling it, the compiler will also mark it as _unchecked throws_. If an error from an _unchecked throws_ function is unhandled, the program will `halt()`.

This will likely be implemented to handle low level errors, but it may be a strange fit for out of bounds errors.

### compiler optimization
proposed by @bradcray 

`boundsChecking` would likely gate the only `throw` in an array accessor. Hence if `boundsChecking` is false, it may be possible to delete that entire conditional and mark the function as non-throwing, which would in turn remove the conditional check in question.

### zero-cost exceptions
proposed by @vasslitvinov 

[Most JVM implementations have no overhead in non-error cases.](http://www.javaworld.com/article/2076868/learn-java/how-the-java-virtual-machine-handles-exceptions.html) It may be possible to extend this approach to our own error handling.

### throw param
proposed by @benharsh 

Have functions that throw take a param, `throws(bool)`, allowing the caller to opt in or out of throwing an error. Implementation of this approach would have to be done at compile time, or else it will just create another runtime conditional.

### separate functions
proposed by @psahabu

Create two functions gated by `where` clauses on `boundsChecking`. Downside is maintainability, but it may be possible to do some effective refactoring.

**Comment by lydia-duncan** [comment]
I like the compiler optimization and separate function solutions the best

**Comment by mppf** [comment]
I personally think we should stick to status quo until somebody agitates for an improvement in this area. Have any users complained about this, e.g.?

**Comment by bradcray** [comment]
@mppf: Just to make sure I'm understanding: This means users would have no way to catch an OOB error ever, right? (i.e., even if they wanted to).  I don't think the issue with the status quo is that users have complained about it; more that in a Chapel world with error-checking, it arguably seems weird not to have any mode in which such errors can be caught -- i.e., your only choices are halts or memory faults.

**Comment by mppf** [comment]
@bradcray - such errors can currently be caught by explicit logic, such as `myArray.domain.member(i)`.

I view this discussion as similar to the integer overflow discussion we had in issue #6160. But the main reason I'm advocating for *status quo* on this issue is that the current behavior is philosophically reasonable and that doing more than that is:

  1. possibly unnecessary work
  2. a distraction from the large number of other error handling tasks that I view as more important.

Why is the status quo (halt on OOB access with checks/without --fast; undefined behavior with --fast) reasonable?
 * I don't think it's likely that we'll choose to perform array bounds checking with --fast
 * Therefore any bounds checking `throw` won't happen with --fast
 * Therefore a program `catch`ing OOB errors would not function correctly with --fast.
 * I think this would cause confusion in a particularly unpleasant way. Users might think they had a safety net, but then try to run their program in production or with a big data set, only to find it crash (or worse). I'd rather that we more straightforwardly indicate that an OOB array access is a fatal error that is on the programmer to fix.
 * Another way to put it is that I don't think that --fast is supposed to change the behavior of correct programs. And I don't think we want to guarantee the behavior of incorrect programs. And I especially don't think we should guarantee particular behavior of incorrect programs depending on optimization level.

In the terms of the overflow issue #6160, I think that a OOB array access occurs only in an incorrect program - that is, it causes *undefined behavior*. I think it's reasonable to go from *undefined behavior* to *trap* in debug / `--checks` / not fast mode. I don't have a problem with that because it's still clear to the programmer that it's a serious error (i.e. just wrapping it in `try`/`catch` isn't an available solution). I expect this will lead the programmer to find a suitable solution (usually, fixing a bug; sometimes, adding an explicitly check with `domain.member`).

I think we should reserve `throw`/`catch`/etc for errors that are detected whether or not `--fast` is thrown. Errors that are not detected when `--fast` is active should be *undefined behavior* as far as the language definition is concerned (but I don't think it's too confusing to give them a more friendly error message. It's just that programs should not be able to handle this error).

**Comment by vasslitvinov** [comment]
I support keeping bounds checking as-is for now because we have other priorities for this release.

For the future, I like the philosophy "OOB programs are incorrect, we should not cater to them". However, here are a couple of considerations:

* Thinking of nil dereference as being in the same category, do we want to support this style?

```chpl
try {
  writeln(me.I.love[3].dereferencing[45].a.bunch.of.fields);
} catch {
  // oops that did not work, I better do something else
}
```

* We want to provide an error-throwing array indexing operation. For example, if indexing involves a list traversal, without such a combo the users would need to do it twice, first to check if index is there, then again to actually get at it.

**Comment by mppf** [comment]
The current consensus direction is to keep bounds checking as-is for now (doesn't interact directly with error handling).

**Comment by mppf** [comment]
I'm closing this issue since, at least for this release, we did *status quo*. If we change our mind or identify further work, we can reopen it or create a new issue.
